### PR TITLE
Upgrade Whitehall to Rails 4.2

### PR DESCRIPTION
**Pull request by boffbowsh**
Quite a mammoth Pull Request. Hopefully the commits tell a reasonable story on this, due to the number of changes it's probably best to review this commit-by-commit.

Main breaking changes that needed addressing:

- Changes to internal API for Routing, used for the localised switch
- Change to internal reflection/association API now using strings instead of symbols
- `assert_select` now provided by the `rails-dom-testing` gem, which uses Nokogiri under the hood
- Rails being stricter about needing ActiveModel compatibility everywhere, namely `model_name` and `to_model`
- `schema.rb` has some new things, like field length limits and more options for `:force`. This has been auto-generated, so if it fails blame Rails not me.
- New common API to typecast from db types to Ruby types, used for booleans
- Seems like Strong Parameters is now stricter, somehow. Had to add a bunch more explicit parameters.
- ActionMailer now has `deliver_now` and `deliver_later`, I guess to distinguish for ActiveJob
- Can't call `#find` or `#exists?` with AR instances any more, only with the id. This was mostly used in tests
- If using MySQL 5.6 or above like I am, Rails 4.2 introduces nanosecond resolution on timestamps. I fixed this in one test for active editors of versions, but it does exist in other places. Not really needed for right now, but something to be aware of if/when we upgrade to 5.6
- Changes to the way path globbing works in router. I haven't pinned this down exactly, but it seems to be around automatically consuming the :format from the route, even if it's not specified in the path spec
- Should use secrets.yml for the secret keys. May need deployment changes?
- Upgraded `friendly_id` alongside Rails. It removed the override of `.find` and added a `.friendly` scope, so `Organisation.find(slug) becomes `Organisation.friendly.find(slug)`.

Todo:
- [x] Random test failure, dependent on ordering of tests. Tests are now run in a random order, and is causing a very tricky failure in the after_create callback for Editions.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -54,13 +55,13 @@ else
   # globalize master, but that version is only compatible with ActiveRecord 4.2
```
Comment (jamiecobbett) [inline Gemfile]: Seems like this comment can go now?

File: app/controllers/admin/attachments_controller.rb [hunk 2]
```diff
@@ -145,7 +145,10 @@ def attachable_id
   end
 
   def attachable
-    @attachable ||= attachable_class.find(attachable_id)
+    @attachable ||= begin
+      scope = attachable_class.respond_to?(:friendly) ? attachable_class.friendly : attachable_class
```
Comment (evilstreak) [inline app/controllers/admin/attachments_controller.rb]: This feels like it ought to be in a different method, eg `attachable_scope`

File: Gemfile [hunk 3]
```diff
@@ -69,6 +65,12 @@ else
   gem 'govuk_frontend_toolkit', '3.1.0'
 end
 
+if ENV['SLIMMER_DEV']
+  gem 'slimmer', path: '../slimmer'
+else
+  gem 'slimmer', '8.1.0', github: "alphagov/slimmer", branch: "obey-strict-mode"
```
Comment (evilstreak) [inline Gemfile]: A new version of slimmer will need to be released with the change we want to be used here before this PR can be merged.

**Comment by evilstreak** [comment]
It's good to see the commits with this granularity, it makes it easy to follow. That said, I feel like a lot of these commits could benefit from more explanation of why the change is being made or why the change is necessary.

**Comment by boffbowsh** [comment]
Reworded some commits to add additional context

File: Gemfile (hunk 3 above)
Comment (boffbowsh) [inline Gemfile]: Removed this.

File: app/controllers/admin/attachments_controller.rb (hunk 2 above)
Comment (boffbowsh) [inline app/controllers/admin/attachments_controller.rb]: Done.

**Comment by evilstreak** [comment]
:+1: Nice work.

**Comment by heathd** [comment]
:+1: nice work

**Pull request merged**